pkg/servicescm/servicescm.go

```diff
@@ -125,6 +134,152 @@ func Generate(name, namespace string) (*core.ConfigMap, error) {
 	return servicesConfigMap, nil
 }
 
+// Parse converts ConfigMap data into the objects representing a Windows services ConfigMap schema
+// Returns error if the given data is invalid in structure or contents.
+func Parse(dataFromCM map[string]string) (*data, error) {
+	if len(dataFromCM) != 2 {
+		return nil, errors.New("services ConfigMap should have exactly 2 keys")
+	}
+
+	value, ok := dataFromCM[servicesKey]
+	if !ok {
+		return nil, errors.Errorf("expected key %s does not exist", servicesKey)
+	}
+	services := &[]Service{}
+	if err := json.Unmarshal([]byte(value), services); err != nil {
+		return nil, err
+	}
+
+	value, ok = dataFromCM[filesKey]
+	if !ok {
+		return nil, errors.Errorf("expected key %s does not exist", filesKey)
+	}
+	files := &[]FileInfo{}
+	if err := json.Unmarshal([]byte(value), files); err != nil {
+		return nil, err
+	}
+
+	return newData(services, files)
+}
+
+// validate ensures the given object represents a valid services ConfigMap, made up of only required services and files.
+// The validation also ensures bootstrap services are defined to always start before controller services.
+func (cmData *data) validate() error {
+	if err := validateRequiredContent(cmData.Services, cmData.Files); err != nil {
+		return err
+	}
+	if err := validateDependencies(cmData.Services); err != nil {
+		return err
+	}
+	return validatePriorities(cmData.Services)
+}
+
+// validateRequiredContent ensures that the given slices are comprised of all the required services/files and only these
+func validateRequiredContent(services []Service, files []FileInfo) error {
+	// Validate services
+	if len(services) != len(*requiredServices) {
+		return errors.New("Unexpected number of services")
+	}
+	for _, requiredSvc := range *requiredServices {
+		if !requiredSvc.isPresent(services) {
+			return errors.Errorf("Required service %s is not present with expected configuration", requiredSvc.Name)
+		}
+	}
+	// Validate files
+	if len(files) != len(*requiredFiles) {
+		return errors.New("Unexpected number of files")
+	}
+	for _, requiredFile := range *requiredFiles {
+		if !requiredFile.isPresent(files) {
+			return errors.Errorf("Required file %s is not present as expected", requiredFile.Path)
+		}
+	}
+	return nil
+}
+
+// isPresent checks if the required service exists within the given services slice
+func (s Service) isPresent(services []Service) bool {
+	for _, service := range services {
+		if reflect.DeepEqual(s, service) {
+			return true
+		}
+	}
+	return false
+}
+
+// isPresent checks if the required file exists within the given files slice
+// TODO: When we move to go1.18, consolodate this helper with the above Service.isPresent using generics
```

Comment:
Why do we have to wait until 1.18 to consolidate?

Reply:
Only 1.18+ supports generic types